marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs

```diff
@@ -0,0 +1,178 @@
+        getValueAtAddress targetAddress txos = getFirst . mconcat . map (First
+          . (\(C.TxOut addressInEra txOutValue _ _) ->
+              if fromCardanoAddressInEra C.BabbageEra addressInEra == targetAddress
+                then Just txOutValue
+                else Nothing))
+          $ txos
+
+      txBodyContent <- hedgehog $ genTxBodyContent C.BabbageEra
+
+      let
+        actual = (getValueAtAddress marloweAddress . C.txOuts)
+          <$> Constraints.adjustTxForMinUtxo protocol marloweAddress txBodyContent
+        expected :: Either (Constraints.ConstraintError 'Core.Api.V1) (Maybe (C.TxOutValue C.BabbageEra))
+          = Right $ getValueAtAddress marloweAddress $ C.txOuts txBodyContent
+
+      pure $ actual `shouldBe` expected
+
+    Hspec.QuickCheck.prop "All outputs satisfy Cardano API minimum UTxO requirements" do
+      marloweScriptHash <- hedgehog genScriptHash
+
+      let
+        scriptAddress hash = fromCardanoAddressAny
+          $ C.AddressShelley
+          $ C.makeShelleyAddress C.Mainnet (C.PaymentCredentialByScript hash) C.NoStakeAddress
+        marloweAddress = scriptAddress marloweScriptHash
+
+        valueMeetsMinimumReq :: C.TxOut C.CtxTx C.BabbageEra -> Bool
+        valueMeetsMinimumReq txout@(C.TxOut _ txOrigValue _ _) =
+          case C.calculateMinimumUTxO C.ShelleyBasedEraBabbage txout protocolTestnet of
+            Right minValueFromApi ->
+              (C.selectLovelace . C.txOutValueToValue $ txOrigValue)
+              >= (C.selectLovelace minValueFromApi)
+            Left _ -> False
+
+      txBodyContent <- hedgehog $ genTxBodyContent C.BabbageEra
+
+      pure $ case Constraints.adjustTxForMinUtxo protocolTestnet marloweAddress txBodyContent of
+        Right newTxBodyContent ->
+          all (== True) (map valueMeetsMinimumReq $ C.txOuts newTxBodyContent) `shouldBe` True
+        Left _ -> expectationFailure "Minimum UTxO requirements not met"
+
+    Hspec.QuickCheck.prop "All outputs are at least half an ADA" do
+      marloweScriptHash <- hedgehog genScriptHash
+
+      let
+        scriptAddress hash = fromCardanoAddressAny
+          $ C.AddressShelley
+          $ C.makeShelleyAddress C.Mainnet (C.PaymentCredentialByScript hash) C.NoStakeAddress
+        marloweAddress = scriptAddress marloweScriptHash
+
+        valueIsAtLeastHalfAnAda :: C.TxOut C.CtxTx C.BabbageEra -> Bool
+        valueIsAtLeastHalfAnAda (C.TxOut _ txOrigValue _ _) =
+          (C.selectLovelace . C.txOutValueToValue $ txOrigValue) >= C.Lovelace 500_000
+
+      txBodyContent <- hedgehog $ genTxBodyContent C.BabbageEra
+
+      pure $ case Constraints.adjustTxForMinUtxo protocolTestnet marloweAddress txBodyContent of
+        Right newTxBodyContent ->
+          all (== True) (map valueIsAtLeastHalfAnAda $ C.txOuts newTxBodyContent) `shouldBe` True
+        Left _ -> expectationFailure "Minimum UTxO requirements not met"
```

Comment:
Same suggestion
```suggestion
      pure $ case Constraints.adjustTxForMinUtxo protocolTestnet marloweAddress txBodyContent of
        Right newTxBodyContent ->
          counterexample
            (show $ C.txOuts newTxBodyContent)
            (all (== True) (map valueIsAtLeastHalfAnAda $ C.txOuts newTxBodyContent) `shouldBe` True)
        Left err -> counterexample (show err) False
```

Reply:
Included in the fix for the prior comment.